chips/rp2040/src/i2c.rs

```diff
@@ -0,0 +1,588 @@
+        match self.instance_num {
+            0 => resets.reset(&[resets::Peripheral::I2c0]),
+            1 => resets.reset(&[resets::Peripheral::I2c1]),
+            _ => unreachable!(),
+        };
+    }
+
+    fn unreset(&self, resets: &'static resets::Resets) {
+        match self.instance_num {
+            0 => resets.unreset(&[resets::Peripheral::I2c0], true),
+            1 => resets.unreset(&[resets::Peripheral::I2c1], true),
+            _ => unreachable!(),
+        };
+    }
+
+    fn disable(&self) {
+        self.registers.ic_enable.set(0);
+    }
+
+    fn enable(&self) {
+        self.registers.ic_enable.modify(IC_ENABLE::ENABLE::SET);
+    }
+
+    fn set_baudrate(&self, baudrate: u32, clocks: &'static clocks::Clocks) -> u32 {
+        assert!(baudrate != 0);
+
+        // I2C is synchronous design that runs from clk_sys
+        let freq_in = clocks.get_frequency(clocks::Clock::System);
+
+        // TODO: as per the comments in the pico-sdk, this block is not 100% correct
+        let period = (freq_in + baudrate / 2) / baudrate;
+        let lcnt = period * 3 / 5;
+        let hcnt = period - lcnt;
+        assert!(hcnt >= 8);
+        assert!(lcnt >= 8);
+
+        // Per I2C-bus specification a device in standard or fast mode must
+        // internally provide a hold time of at least 300ns for the SDA signal to
+        // bridge the undefined region of the falling edge of SCL. A smaller hold
+        // time of 120ns is used for fast mode plus.
+        let sda_tx_hold_count;
+        if baudrate < 1000000 {
+            // sda_tx_hold_count = freq_in [cycles/s] * 300ns * (1s / 1e9ns)
+            // Reduce 300/1e9 to 3/1e7 to avoid numbers that don't fit in uint.
+            // Add 1 to avoid division truncation.
+            sda_tx_hold_count = ((freq_in * 3) / 10000000) + 1;
+        } else {
+            // sda_tx_hold_count = freq_in [cycles/s] * 120ns * (1s / 1e9ns)
+            // Reduce 120/1e9 to 3/25e6 to avoid numbers that don't fit in uint.
+            // Add 1 to avoid division truncation.
+            sda_tx_hold_count = ((freq_in * 3) / 25000000) + 1;
+        }
+        assert!(sda_tx_hold_count <= lcnt - 2);
+
+        self.registers.ic_enable.modify(IC_ENABLE::ENABLE::CLEAR);
+        // Always use "fast" mode (<= 400 kHz, works fine for standard mode too)
+        self.registers.ic_con.modify(IC_CON::SPEED::FAST);
+        self.registers.ic_fs_scl_hcnt.set(hcnt);
+        self.registers.ic_fs_scl_lcnt.set(lcnt);
+        self.registers.ic_fs_spklen.set({
+            if lcnt < 16 {
+                1
+            } else {
+                lcnt / 16
+            }
+        });
+        self.registers
+            .ic_sda_hold
+            .modify(IC_SDA_HOLD::IC_SDA_TX_HOLD.val(sda_tx_hold_count));
+
+        freq_in / period
+    }
+
+    pub fn init(
+        &self,
+        baudrate: u32,
+        clocks: &'static clocks::Clocks,
+        resets: &'static resets::Resets,
```

Comment:
Can you add the `reset` and `clocks` as parameters to the I2C structure, similar to `clocks` for SPI (https://github.com/tock/tock/blob/9f159e0c06e27b2357e3b20be6c90b3396bbd7b9/chips/rp2040/src/spi.rs#L254)

Reply:
I started this patch before SPI was merged and finished before noticing, but interestingly I actually started out doing exactly what you suggest here. I removed it in a refactoring pass because the references are not needed again after init() and I felt it was a waste of space to keep them stored in the struct for no good reason. Is there a technical benefit to storing it or do we just want to keep SPI and I2C more stylistically similar?

Reply:
Looks good now.